packages/core/core/src/RequestTracker.js

```diff
@@ -313,15 +353,76 @@ export class RequestGraph extends Graph<
     }
   }
 
-  invalidateOnFileCreate(requestId: string, glob: Glob) {
+  invalidateOnFileCreate(requestId: string, input: FileCreateInvalidation) {
     let requestNode = this.getRequestNode(requestId);
-    let globNode = nodeFromGlob(glob);
-    if (!this.hasNode(globNode.id)) {
-      this.addNode(globNode);
+    let node;
+    if (input.glob != null) {
+      node = nodeFromGlob(input.glob);
+    } else if (input.fileName != null && input.aboveFilePath != null) {
+      let aboveFilePath = input.aboveFilePath;
+
+      // Create nodes and edges for each part of the filename pattern.
+      // For example, 'node_modules/foo' would create two nodes and one edge.
+      // This creates a sort of trie structure within the graph that can be
+      // quickly matched by following the edges. This is also memory efficient
+      // since common sub-paths (e.g. 'node_modules') are deduplicated.
+      let parts = input.fileName.split('/').reverse();
+      let last;
+      for (let part of parts) {
+        let fileNameNode = nodeFromFileName(part);
+        if (!this.hasNode(fileNameNode.id)) {
+          this.addNode(fileNameNode);
+        }
+
+        if (
+          last != null &&
+          !this.hasEdge(last.id, fileNameNode.id, 'dirname')
+        ) {
+          this.addEdge(last.id, fileNameNode.id, 'dirname');
+        }
+
+        last = fileNameNode;
+      }
+
+      // The `aboveFilePath` condition asserts that requests are only invalidated
+      // if the file being created is "above" it in the filesystem (e.g. the file
+      // is created in a parent directory). There is likely to already be a node
+      // for this file in the graph (e.g. the source file) that we can reuse for this.
+      node = nodeFromFilePath(aboveFilePath);
+      if (!this.hasNode(node.id)) {
+        this.addNode(node);
+      }
+
+      // Now create an edge from the `aboveFilePath` node to the first file_name node
+      // in the chain created above, and an edge from the last node in the chain back to
+      // the `aboveFilePath` node. When matching, we will start from the first node in
+      // the chain, and continue following it to parent directories until there is an
+      // edge pointing an `aboveFilePath` node that also points to the start of the chain.
+      // This indicates a complete match, and any requests attached to the `aboveFilePath`
+      // node will be invalidated.
+      let firstId = 'file_name:' + parts[0];
+      if (!this.hasEdge(node.id, firstId, 'invalidated_by_create_above')) {
+        this.addEdge(node.id, firstId, 'invalidated_by_create_above');
+      }
+
+      if (
+        last != null &&
```

Comment:
Could `last` actually be null in some situations? Might be safer to do a nullthrows